arches/management/commands/validate.py

```diff
@@ -0,0 +1,100 @@
+"""
+ARCHES - a program developed to inventory and manage immovable cultural heritage.
+Copyright (C) 2013 J. Paul Getty Trust and World Monuments Fund
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as
+published by the Free Software Foundation, either version 3 of the
+License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+"""
+
+from datetime import datetime
+
+from arches import __version__
+from arches.app.const import IntegrityCheck
+from arches.app.models import models
+from django.core.management.base import BaseCommand
+from django.db import transaction
+from django.db.models import Exists, OuterRef
+
+DELETE_QUERYSET = "delete queryset"
+
+
+class Command(BaseCommand):
+    """
+    Validate an Arches database against a set of data integrity checks.
+    Takes no action by default (other than printing a summary).
+
+    Provide --verbosity=2 to get a richer output (list of affected nodes).
+    """
+    help = "Validate an Arches database against a set of data integrity checks."
+
+    def add_arguments(self, parser):
+        parser.add_argument("--fix-all", action="store_true", dest="fix_all", default=False,
+                            help="Apply all fix actions.")
+        parser.add_argument("--fix", action="extend", nargs="+", type=int, default=[],
+                            help="List the error codes to fix, e.g. --fix 1001 1002 ...")
+        parser.add_argument("--limit", action="store", type=int, default=500,
+                            help="Maximum number of rows to print; does not affect fix actions")
+
+    def handle(self, *args, **options):
+        self.options = options
+        if self.options["verbosity"] > 0:
+            self.stdout.write()
+            self.stdout.write("Arches integrity report")
+            self.stdout.write(f"Prepared by Arches {__version__} on {datetime.today()}")
+            self.stdout.write()
+            self.stdout.write("\t".join(["", "Error", "Rows", "Description"]))
+            self.stdout.write()
+        
+        # add checks here in numerical order
+        self.check_integrity(
```

Comment:
Maybe we need another column in the output specifiying if there is a fix available.  In this case there isn't.

Reply:
I think what I'd like to see is something more specific.  When running in just validate mode I'd love to see an output like this.

```
(env_7x) alexeis-16-mbp:accessibility alexei$ python manage.py validate --limit=1

Arches integrity report
Prepared by Arches 7.4.1b0 on Fri Aug 11 14:06:51 2023

        Error   Rows    Fixable?        Description

FAIL    1005    1       No              Nodes with ontologies found in graphs without ontologies
        ------------------------------------
        0e8ff9d1-4148-11e7-90ed-c4b301baab9f

FAIL    1012    1       Yes             Node Groups without matching nodes
        ------------------------------------
        0e8ff9d1-4148-11e7-90ed-c4b301baab9f
```

Running in "fix" mode would give you a slightly different output
```
(env_7x) alexeis-16-mbp:accessibility alexei$ python manage.py validate --fix-all

Arches integrity report
Prepared by Arches 7.4.1b0 on Fri Aug 11 14:06:51 2023

        Error   Rows    Fixed?        Description

FAIL    1005    1       N/A           Nodes with ontologies found in graphs without ontologies

PASS    1012    0       Yes           Node Groups without matching nodes

```
Finally, running only a single fix (ie: --fix 1012) would only output the results of that fix
```
(env_7x) alexeis-16-mbp:accessibility alexei$ python manage.py validate --fix 1012

Arches integrity report
Prepared by Arches 7.4.1b0 on Fri Aug 11 14:06:51 2023

        Error   Rows    Fixed?        Description

PASS    1012    0       Yes           Node Groups without matching nodes

```